x-pack/plugins/discover_log_explorer/common/datasets/models/dataset.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { DatasetId, DatasetType, IntegrationType } from '../types';
+
+type IntegrationBase = Pick<IntegrationType, 'name' | 'version'>;
+
+interface DatasetDeps {
+  dataset: DatasetType;
+  integration?: IntegrationBase;
+}
+
+interface DatasetSpec {
+  id: DatasetId;
+  name: DatasetType['title'];
+  title: DatasetType['name'];
+}
+
+export interface DatasetPlain {
+  dataset: DatasetType & { id: DatasetId };
+  integration?: IntegrationBase;
+}
+
+export interface IDataset {
+  id: DatasetId;
+  name: DatasetType['name'];
+  title: DatasetType['title'];
+  integration?: IntegrationBase;
+  toPlain: () => DatasetPlain;
+  toSpec: () => DatasetSpec;
+}
+
+export class Dataset implements IDataset {
```

Comment:
Can you help me understand what these model classes offer over just a `Dataset` interface (that is equivalent to `DatasetPlain`) and a `createDataviewSpecFromDataset()` function?

Reply:
Sure, the approach you are proposing makes sense as it keeps the Integration and Dataset usage simple, I opted for using classes (with the overhead they bring) for the following reasons:
- I want to create the unique id associated with each integration and dataset only once the data is fetched (could be done with a factory function equivalently).
- I want to associate, if it has one, the parent integration with a Dataset, for easy use when using the selected dataset. I also needed to build the ~presentational title composed as `[<integration-name>] <dataset-name>`~, now I use it to resolve the integration icon.
- I want to centralize the fallback to a Dataset name when a title is not provided (could be done with the same factory function used above)
- I want to encapsulate the derivation of Data view specs starting from a dataset and its available data.

All of the above changes are minimal and can be achieved with functions and interfaces, however, I preferred encapsulating that logic inside model classes in prevision of the upcoming work, where we'll need to have a way to maybe search for a dataset inside an integration, know if any dataset of integration is selected etc... so I thought having those methods contextually bound to each instance would be easier to consume in the long run.

I updated the Dataset implementation to simplify its creation and reduce the different types.

Let me know if you see any downside to this approach, I'm happy to talk about it and apply any suggestions to improve this code :)

Reply:
I can understand those reasons, thanks for the explanation. This just surprised me because it's not a pattern I've seen before in this code-base. The only downside I can see is that it's not serializable as such, but at the same time I don't see a need for that at the moment.